### PR TITLE
Rework PIT example train.py and data.py

### PR DESCRIPTION
**Pull request by sibange**

### CODE REVIEW

File: padertorch/contrib/examples/source_separation/pit/data.py [hunk 1]
```diff
@@ -8,25 +6,34 @@
 from paderbox.transform import stft
 
 
-def prepare_dataset(
-        db, dataset_name: str, batch_size, return_keys=None, prefetch=True, shuffle=True
-):
-    audio_keys = ['observation', 'speech_source']
+def prepare_dataset(db, dataset_name: str, batch_size, prefetch=True, shuffle=True):
+    """
+    Prepares the dataset for the training process (loading audio data, SFTF)
```
Comment (thequilo) [inline padertorch/contrib/examples/source_separation/pit/data.py:11]: Typo: SFTF -> STFT

File: padertorch/contrib/examples/source_separation/pit/data.py [hunk 2]
```diff
@@ -8,25 +6,34 @@
...
+
+    Args:
+        db: database object
+        dataset_name: name of the dataset that should be used
+        batch_size: batch size for the training
+        prefetch: should the data be prefetched
+        shuffle: should the data be shuffeled
+
+    Returns:
+        desired dataset of the database in prepared for the training
```
Comment (thequilo) [inline padertorch/contrib/examples/source_separation/pit/data.py:21]: Something is wrong with the grammar in this sentence

File: padertorch/contrib/examples/source_separation/pit/train.py [hunk 3]
```diff
@@ -79,104 +81,141 @@ def config():
...
-    """ Creates a storage dir, writes Makefile. Does not start any training."""
+    """
+    Creates/updates config file (config.json) with current configurations of the pit experiment.
+    Writes Makefile if no Makefile currently exists. (e.g. when train.py is called the first time.)
+
+    Args:
+        _config: Configuration dict of the experiment
+        _run: Run object of the current run of the experiment
+
+    Returns:
```
Comment (thequilo) [inline padertorch/contrib/examples/source_separation/pit/train.py:94]: This can be left out when there is not return value

File: padertorch/contrib/examples/source_separation/pit/train.py [hunk 4]
```diff
@@ -79,104 +81,141 @@ def config():
...
 
-    trainer.test_run(
-        prepare_dataset_captured(db, train_dataset),
-        prepare_dataset_captured(db, validate_dataset),
-    )
+    Returns:
+        None
+    """
+    init(_config, _run)
+    (trainer, train_dataset, validate_dataset) = prepare(_config)
```
Comment (thequilo) [inline padertorch/contrib/examples/source_separation/pit/train.py:192]: The parentheses on the left-hand-side are redundant

File: padertorch/contrib/examples/source_separation/pit/train.py [hunk 5]
```diff
@@ -79,104 +81,141 @@ def config():
...
+        validate_dataset: dataset for validation
+
+    Returns:
+        None
+    """
+
+    # Test run to detects possible errors in the trainer/datasets
+    trainer.test_run(train_dataset, validate_dataset)
+
+    # path where the checkpoints of the training are stored
```
Comment (thequilo) [inline padertorch/contrib/examples/source_separation/pit/train.py:172]: This comment is lower case, others are upper case. Stick to one (I prefer upper case)

File: padertorch/contrib/examples/source_separation/pit/data.py [hunk 6]
```diff
@@ -8,25 +6,34 @@
...
-        .map(partial(read_audio, audio_keys=audio_keys))
-        .map(partial(pre_batch_transform, return_keys=return_keys))
-    )
+    #Loading of the dataset and preparation of the data (STFT), the lambda function fixes the audio keys for read_audio
+    dataset = (dataset.map(read_audio).map(pre_batch_transform))
+
     if shuffle:
         dataset = dataset.shuffle(reshuffle=True)
+
+    #Splitting the dataset in batches and sorting the frames in the batch
```
Comment (TCord) [inline padertorch/contrib/examples/source_separation/pit/data.py:31]: Better write "... and sorts examples in a batch w.r.t. their duration" or something similar.
The frames themselves are not sorted

File: padertorch/contrib/examples/source_separation/pit/data.py [hunk 7]
```diff
@@ -35,49 +42,57 @@ def prepare_dataset(
...
     }
-    example["audio_data"] = data
+    example['audio_data'] = data
     return example
 
 
-def pre_batch_transform(inputs, return_keys=None):
+def pre_batch_transform(inputs):
+    """
+    Prepares the data through creating a dictionary with various data, which is computed through STFT.
```
Comment (TCord) [inline padertorch/contrib/examples/source_separation/pit/data.py:71]: "... by creating a dictionary with all data that is necessary for the model (e.g. STFT of observation)"

File: padertorch/contrib/examples/source_separation/pit/data.py [hunk 8]
```diff
@@ -35,49 +42,57 @@ def prepare_dataset(
...
         for audio_key in audio_keys
     }
-    example["audio_data"] = data
+    example['audio_data'] = data
     return example
 
 
-def pre_batch_transform(inputs, return_keys=None):
+def pre_batch_transform(inputs):
+    """
```
Comment (TCord) [inline padertorch/contrib/examples/source_separation/pit/data.py:70]: Could you also explain our nomenclature in the doc string?
s: speech source
x: speech image
y: observation
Capital letters: STFT of signal

File: padertorch/contrib/examples/source_separation/pit/train.py [hunk 9]
```diff
@@ -79,104 +81,141 @@ def config():
...
-    print(f"cd {_config['trainer']['storage_dir']}")
-    print('make ccsalloc')
 
 
-@ex.capture
-def prepare_and_train(_config, _run, train_dataset, validate_dataset, database_json):
-    """ Prepares the train and validation dataset from the database object """
+def prepare(_config):
+    """
+    Preparation of the train and validation datasets for the training and initialisation of the padertorch trainer,
```
Comment (TCord) [inline padertorch/contrib/examples/source_separation/pit/train.py:118]: We try to stick to American English. intitialisation -> initialization

File: padertorch/contrib/examples/source_separation/pit/train.py [hunk 10]
```diff
@@ -79,104 +81,141 @@ def config():
...
+        validate_dataset: dataset for validation
+    """
+
+    # Extraction needed strings from the config dict
+    train_dataset_name = _config['train_dataset_name']
+    validate_dataset_name = _config['validate_dataset_name']
+    database_json = _config['database_json']
 
-    sacred.commands.print_config(_run)
+    # Initialisation of the trainer
```
Comment (TCord) [inline padertorch/contrib/examples/source_separation/pit/train.py:135]: intitialisation -> initialization

File: padertorch/contrib/examples/source_separation/pit/train.py [hunk 11]
```diff
@@ -79,104 +81,141 @@ def config():
...
+    """
+
+    # Test run to detects possible errors in the trainer/datasets
+    trainer.test_run(train_dataset, validate_dataset)
+
+    # path where the checkpoints of the training are stored
+    checkpoint_path = trainer.checkpoint_dir / 'ckpt_latest.pth'
+
+    # Start of the training
+    trainer.register_validation_hook(validate_dataset)
```
Comment (TCord) [inline padertorch/contrib/examples/source_separation/pit/train.py:176]: Could you repeat the most important default arguments of the validation hook, so that it becomes clear, what options can be easily modified for the validation (number of checkpoints, metric for the best checkpoint, ...)

**Comment by thequilo** [comment]
Now that the `maybe_add` function disappeared from the data preparation, can you add an `example_to_device` method to the model that only transfers to the GPU those keys from the example that are required for training? You can use `pt.data.batch.example_to_device` for that
